src/targets/maven.ts

```diff
@@ -187,31 +193,140 @@ export class MavenTarget extends BaseTarget {
    * @param revision Git commit SHA to be published.
    */
   public async publish(_version: string, revison: string): Promise<void> {
-    await this.createUserGradlePropsFile();
-    await this.upload(revison);
+    await this.withGradleProps(async () => {
+      await this.upload(revison);
+
+      // Maven central is very flaky, so retrying with an exponential delay in
+      // in case it fails.
+      // TODO: close the repository by doing the requests from Craft
+      // What the gradle plugin does is a / some requests to close the repo, see
+      // https://github.com/vanniktech/gradle-maven-publish-plugin/tree/master/src/main/kotlin/com/vanniktech/maven/publish/nexus
+      // If Craft did those requests, it wouldn't be necessary to rely on a third
+      // party plugin for releases.
+      await retrySpawnProcess(this.mavenConfig.gradleCliPath, [
+        'closeAndReleaseRepository',
+      ]);
+    });
+  }
 
-    // Maven central is very flaky, so retrying with an exponential delay in
-    // in case it fails.
-    // TODO: close the repository by doing the requests from Craft
-    // What the gradle plugin does is a / some requests to close the repo, see
-    // https://github.com/vanniktech/gradle-maven-publish-plugin/tree/master/src/main/kotlin/com/vanniktech/maven/publish/nexus
-    // If Craft did those requests, it wouldn't be necessary to rely on a third
-    // party plugin for releases.
-    await retrySpawnProcess(this.mavenConfig.gradleCliPath, [
-      'closeAndReleaseRepository',
-    ]);
-    await this.deleteUserGradlePropsFile();
+  /**
+   * Encapsulates the given callback in a context manager. Maven CLI requires
+   * Maven Central credentials to be available in the gradle properties, which
+   * should be deleted to not expose them. The context manager makes a snapshot
+   * of the gradle properties, adds the credentials, runs the callback, and
+   * restores the snapshot.
+   *
+   * @param cb Callback to encapsulate.
+   */
+  private async withGradleProps(
+    cb: CallbackFunction
+  ): ReturnType<CallbackFunction> {
+    await withTempDir(async dir => {
+      const snapshotPath = await this.makeGradlePropsSnapshot(dir);
+      // Overwrite the file, instead of appending the content, because:
+      // 1. The target must get all the config by itself, without relying on
+      // local config files.
+      // 2. Appending more config to the gradle config file may imply a
+      // misconfiguration (such as duplicated or conflicting keys).
+      await this.createUserGradlePropsFile();
+
+      try {
+        return await cb();
+      } finally {
+        if (snapshotPath) {
+          await this.restoreGradleProps(snapshotPath);
+        } else {
+          this.logger.debug('Deleting temporary gradle properties file...');
+          this.deleteUserGradlePropsFile();
```

Comment:
This logic already exists in `restoreGradleProps()` so I think you should:

1. Remove the conditional here and always call `await this.restoreGradleProps(snapshotPath)`
2. Remove the helper function `deleteUserGradlePropsFile` and bake its code into `restoreGradleProps`